### ISSUE TITLE
Autofix S101

### ISSUE THREAD

**Issue by kunaltyagi** [astral-sh/ruff#7635]
code: `assert x != 0`
ruff: 0.0.291
command: `ruff check --fix --fixable S101 .`
error: `S101 Use of `assert` detected`

This is an easy fix to substitute in the code specially when opted-in by the developer (if not automatically)

**Comment by charliermarsh** [comment]
What would you suggest as the correct fix?

**Comment by kunaltyagi** [comment]
If the assert doesn't have a message, a generic message can be used alongside the assert condition:
```py
if x == 0:
  msg = f"Expected: 'x != 0', found: x = {x}"
  raise ValueError(msg)
```

If there's a message accompanying the assert:
```py
if x == 0:
  msg = <same message as the assert>
  raise ValueError(msg)
```

This is what I've observed people replace their code with when faced with this error. The most often item people might want to change is the error type but this default would more than suffice the 80-20 rule

**Comment by dimbleby** [comment]
`AssertionError` would seem the more natural substitution, as already done when auto-fixing `B011`.

**Comment by kunaltyagi** [comment]
Sure, that does make more sense

**Comment by cosmojg** [comment]
If this is implemented, might it wreak havoc on tests?

**Comment by kunaltyagi** [comment]
I don't think so. Asserts in tests are ignored in the projects I've used ruff with

**Comment by cosmojg** [comment]
Huh, weird! I had to go out of my way to explicitly declare the following in my `pyproject.toml`

```
[tool.ruff.per-file-ignores]
"*test_*.py" = ["S101"]
```

**Comment by charliermarsh** [comment]
We don't disable any specific rules in tests by default, so something like what you have there @cosmojg is expected and correct.

**Comment by cosmojg** [comment]
Ah, got it, thanks for the clarification! That makes sense.

It's an interesting philosophical question whether tests are part of the language itself or merely a product thereof. Regardless, it seems that [bandit](https://bandit.readthedocs.io/en/latest/plugins/b101_assert_used.html) maintains the same behavior upstream. I suppose some users might prefix scripts with "test" to mean that they are tests in some other sense. Also, the Python interpreter itself doesn't differentiate even though the standard library does (see: [unittest](https://docs.python.org/3/library/unittest.html#test-discovery)), and at the end of the day, explicit behavior generally causes fewer problems than implicit behavior.

**Comment by kunaltyagi** [comment]
@cosmojg I found that I was wrong in my guess and all the projects (I work(ed) on) all had the following common items:
```toml
[tool.ruff.per-file-ignores]
# Tests can use magic values, assertions, and relative imports
"tests/**/*" = ["PLR2004", "S101", "TID252"]
```

**Comment by kunaltyagi** [comment]
Has there been any progress on this?

**Comment by collinanderson** [comment]
I agree that this should be a pretty straightforward autofix:

```
assert {condition}, {message}
```

should become:

```
if not ({condition}):
  raise AssertionError({message})
```

If no `{message}` then just:

```
if not ({condition}):
  raise AssertionError
```

(Other rules can then fix it further, like if `not x in y` -> `x not in y`, or `{message}` getting assigned to a variable first, etc.)

I'd suggest starting out as an unsafe-fix in case `pyright` users are blindly fixing all possible issues, and not already using `per-file-ignores` to ignore `S101` on tests.

It could later be made a safe fix after some time for feedback of the unsafe fix.
